### DISCUSSION TITLE
What optimization should I use to increase speed of Flat index? (for several millions of entries?)

### DISCUSSION THREAD

**Discussion by martinenkoEduard** [milvus-io/milvus#18196] [Q&A and General discussion] [answered]
I need to get 100% return rate (Because I am building simple face recognition software) so I suppose I should use flat index because only it can guarantee the search result.

I am going to search by
 FieldSchema(name="embeddings", dtype=DataType.FLOAT_VECTOR, dim=512)
with total amount of enties - about ~7-10 millions.
What are the best ways to increase speed?

**Comment by yhmo** [comment, marked as answer]
Actually Flat index is "no index", brute-force search is applied to entire collection.
If you didn't create index, it would be "Flat".
And if you called drop_index(), it would be "Flat", too.

The search speed of Flat cannot be accelerated.

**Reply by martinenkoEduard** [reply]
So am I Right that 
```

index = {
    "index_type": "FLAT",
    "metric_type": "L2",
    "params":{}
}
hello_milvus.create_index("embeddings", index)
```
Does nothing at all?

How does it behave if there are several fields in a collection?

**Reply by xiaofan-luan** [reply]
add more resources or batch query is so far what you can do if you want 100% accuracy.
Otherwise use HNSW it's fast and accurate

**Reply by yhmo** [reply]
> So am I Right that
> 
> ```
> 
> index = {
>     "index_type": "FLAT",
>     "metric_type": "L2",
>     "params":{}
> }
> hello_milvus.create_index("embeddings", index)
> ```
> 
> Does nothing at all?
> 
> How does it behave if there are several fields in a collection?

"hello_milvus.create_index("embeddings", index)" is v1.x style. v1.x only has one vector field, it doesn't support multiple fields, so this index is for the vector field. Yes, create "FLAT" index, it does nothing at all.

In the v2.x, to create an index, please read the example: https://github.com/milvus-io/pymilvus/blob/master/examples/example_index.py
The v2.x supports multiple fields(currently only allows one vector field and multiple scalar fields). 
The two lines create "IVF_FLAT" index on a vector field named "embedding":
```
index_params = {"index_type": "IVF_FLAT", "metric_type": "L2", "params": {"nlist": 100}}
index = Index(collection, "embedding", index_params)
```

**Reply by martinenkoEduard** [reply]
> Actually Flat index is "no index", brute-force search is applied to entire collection. If you didn't create index, it would be "Flat". And if you called drop_index(), it would be "Flat", too.
> 
> The search speed of Flat cannot be accelerated.

Btw - if I use cluster configuration will it speed up search using FLAT index?
Does it divide data between shards?

**Reply by yhmo** [reply]
> > Actually Flat index is "no index", brute-force search is applied to entire collection. If you didn't create index, it would be "Flat". And if you called drop_index(), it would be "Flat", too.
> > The search speed of Flat cannot be accelerated.
> 
> Btw - if I use cluster configuration will it speed up search using FLAT index? Does it divide data between shards?

Assume there are 10 segments without index in the collection,
- if you deployed only one query node, when search request arrive, the query node will do brute-force search to the 10 segments, one by one.
- if you deployed 2 query nodes, when search request arrive, each query node do brute-force search to 5 segments, the two query nodes can work parallelly. 

So, 2query nodes can get better performance than 1 query node.

**Reply by martinenkoEduard** [reply]
Btw - what if one node goes down by some reason? Does it mean that some data will be unviable for search, but it will still search?

**Reply by yhmo** [reply]
> Btw - what if one node goes down by some reason? Does it mean that some data will be unviable for search, but it will still search?

Milvus can balance segments across query nodes. If a query node shutdown, the segments will be re-balance so that search request is executed correctly.
